maven_iuvs/echelle.py

```diff
@@ -825,3 +966,1289 @@ def find_files_with_geometry(file_index):
+            back_above_i = np.sum(data_cube[i, bg_inds[2]:bg_inds[3]+1, :], axis=0) 
+            backgrounds_oldcal[i, :] = ( back_below_i + back_above_i ) / Nbacks 
+        
+        return backgrounds_oldcal
+        
+        
+def interpolate_CLSF(lambda_c, binedges, CLSF): #
+    """
+    Given a particular lambda_c, this function computes the CLSF as a 
+    function of dlambda = x-lambda_c, where x is the observational data
+    and LSF_x is the x points defined for the instrument LSF. This is because
+    the instrument LSF is defined by delta lambda, the difference from the 
+    central wavelength of the emission. Finally, the computed CLSF is inteprolated
+    at the bin edges. 
+
+    TODO: This function will eventually have to be modified to accept the bin edges from the
+    actual instrument rather than just making it up, but currently the recorded bin widths 
+    are for the standard resolution mode, not the echelle mode. 
+    
+    Parameters
+    ----------
+    lambda_c : float
+               a single wavelength in nm, candidate for a possible central wavelength marking the location of the emission peak.
+    bin_edges : array
+                Defines the bin edges in nm on which the LSF will be interpolated
+    CLSF : array
+           CLSF computed from the LSF in question, result of CLSF_from_LSF.
+
+    Returns
+    ----------
+    interp_CLSF : array
+                  the same CLSF reinterpolated on bin_edges.
+    """
+  
+    # Shift the wavelengths of the bin centers and edges
+    dlambda_binedges = binedges - lambda_c
+    
+    # Ensure that CLSF x is increasing everywhere so interp doesn't have meaningless results
+    if any(np.diff(CLSF) < 0):
+        raise Exception("ValueError: Can't interpolate because x values are not monotonically increasing")
+
+    interp_CLSF = np.interp(dlambda_binedges, CLSF[:, 0], CLSF[:, 1], left=0, right=1) 
+
+    # For some reason, interp function isn't automatically setting the edges to the requested values
+    if any(interp_CLSF > 1):
+        interp_CLSF[np.where(interp_CLSF > 1)] = 1
+
+    return interp_CLSF
+    
+
+def CLSF_from_LSF(LSFx, LSFy):
+    """
+    Compute the empirical CLSF, given some LSF defined by xdata and ydata.
+    
+    Parameters
+    ----------
+    LSFy, LSFx : arrays
+                 x and y points which define the instrument LSF.
+    dx : real (float or int)
+         difference between consecutive x values. If not provided, will be calculated from xdata.
+
+    Returns
+    ----------
+    cumulative : array
+                 The CLSF, valid at the same xdata.
+    """
+
+    # The LSF usually comes in angstroms, and covers ± ~3 Å; put it in nm.
+    if 1 <= np.max(LSFx)<= 4:
+        m = (1/10)
+    elif np.max(LSFx) < 1:
+        m = 1
+
+
+    cumulative = np.zeros((len(LSFx), 2))
+    
+    # Fill the wavelengths...
+    cumulative[:, 0] = LSFx * m
```

Comment:
it would be clearer for this definition to appear directly in the branches of the if statement above rather than setting a value there that's used later

Reply:
Fixed by: f0c60d192fd4cfebf083e2c2a9930d1f48e90bcf